subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java

```diff
@@ -103,21 +105,50 @@ protected DefaultDomainObjectCollection(DefaultDomainObjectCollection<? super T>
     }
 
     public Iterator<T> iterator() {
+        if (constantTimeIsEmpty()) {
+            return Iterators.emptyIterator();
+        }
         return new IteratorImpl(getStore().iterator());
     }
 
     public void all(Action<? super T> action) {
+
         action = whenObjectAdded(action);
 
+        if (constantTimeIsEmpty()) {
+            return;
+        }
+
         // copy in case any actions mutate the store
         // linked list because the underlying store may preserve order
-        Collection<T> copied = new LinkedList<T>(this);
-
-        for (T t : copied) {
-            action.execute(t);
+        // We make best effort not to create an intermediate collection if this container
+        // is empty.
+        Collection<T> copied = null;
+        for (T t : this) {
+            if (copied == null) {
+                copied = Lists.newArrayListWithExpectedSize(estimatedSize());
+            }
+            copied.add(t);
+        }
+        if (copied != null) {
+            for (T t : copied) {
+                action.execute(t);
+            }
         }
     }
 
+    /**
+     * Returns true if, and only if, the store is empty AND we know that we
+     * can query its size in constant time. Otherwise, it returns false, which
+     * doesn't mean that the store is empty, but just that we cannot determine
+     * in constant time.
+     *
+     * @return true if and only if the store is empty and can tell in constant time
+     */
+    private boolean constantTimeIsEmpty() {
```

Comment:
I disagree. It's the same as `estimateSize()`. You wouldn't call that `constantTimeSize()`. The API expresses what the consumer needs (an estimate of size or an estimate of emptiness) and the implementation can do that as efficiently as possible. If constant time is part of the contract, it can be a guarantee in the JavaDoc. 

In my opionion`constantTimeIsEmpty` does not make sense as a verb. It sounds like "check if the constant time is empty".